### PR TITLE
Export Data Source endpoint to support JSON and pagination

### PR DESCRIPTION
**Pull request by SmittieC**
[Ticket](https://dimagi-dev.atlassian.net/browse/SC-2993)
Refer to the ticket for the spec doc

## Product Description

No user facing effects.

### Background
This is preparation work for the DET to support exporting UCRs. 

### The change
This code adds a new API endpoint to v0.5. This will allow the DET to export data source data from the SQL tables. In order to effectively do this, we needed to add an index to the `inserted_at` column in data source tables, so this PR does that also. Simply adding this will show the index diffs for these tables [without running them](https://github.com/dimagi/commcare-hq/blob/0d0de2d87068a2aaf34e3484852ee6fb0c83938d/corehq/apps/userreports/rebuild.py#L143-L146), so I added a new function `get_tables_with_index_diff_only` that returns tables whose only `diff` is this index on the `inserted_at` column. This is then used to exclude these tables from being rebuilt/migrated during deployment. Excluding them will not print the index create statements now.

### Safety story

- Local testing
- Some unit tests
- I tested on Staging. Deployments didn't trigger migrations/rebuilds. I verified this by 1. picking a data source and looking at its table indexes in postgres (None of them had an `inserted_at` index at that time) 2. deploying and 3. checking the datasource table's indexes again. Again, no `inserted_at` index were found for this table. I then 4. rebuilt the datasource after which this index started to appear.

### QA Plan

No QA planned

### Rollback instructions

- [ ] This PR can be reverted after deploy with no further considerations
Things to consider:
- I think the only thing to consider is if any projects are using the newly added API (Those with the EXPORT_DATA_SOURCE_DATA FF enabled)

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/userreports/tests/test_rebuild_migrate.py
Comment (SmittieC) [inline corehq/apps/userreports/tests/test_rebuild_migrate.py:153]: I ran `blue` to format this file (in order to satisfy flake8 so that I can commit). These are just formatting

File: corehq/apps/userreports/tests/test_rebuild_migrate.py
Comment (kaapstorm) [inline corehq/apps/userreports/tests/test_rebuild_migrate.py:161]: Yeah, this is where `blue` and I disagree. I'd have gone with something like ...
```suggestion
        new_date_cols = [c for c in insp.get_columns(table_name) if c['name'] == 'new_date']
        self.assertEqual(len(new_date_cols), 0)
```

**Review by snopoke — COMMENTED** [review]
I haven't reviewed closely but looks good.

File: corehq/apps/api/resources/v0_5.py
Comment (SmittieC) [reply corehq/apps/api/resources/v0_5.py]: I moved some of these function to other files where I thought it might belong. https://github.com/dimagi/commcare-hq/pull/33380/commits/b6cfc2a6eeec0102a6e804e9aa098d80a8d3badf

**Review by calellowitz — COMMENTED** [review]
Left a few comments, and may have followups based on the answers but mostly I have the same question as on the last PR. Has any performance testing been done on this work for large data sources?

**Comment by SmittieC** [comment]
> Left a few comments, and may have followups based on the answers but mostly I have the same question as on the last PR. Has any performance testing been done on this work for large data sources?

No performance testing have been done. Sorry I should have informed reviewers of my plan: I first wanted to get some concerns out of the way that ya'll might have looking at just the code. When this is fine, I'll deploy to staging to do some testing and see how it holds up.

File: corehq/apps/userreports/rebuild.py
Comment (SmittieC) [reply corehq/apps/userreports/rebuild.py:79]: > does this mean the index will not be added to current tables?

yes, if the index on the `inserted_at` column is the only change. It doesn't look like the user can add this index through the UI, so this change shouldn't prevent the user from rebuilding their data source

File: corehq/apps/api/resources/pagination.py
Comment (kaapstorm) [inline corehq/apps/api/resources/pagination.py:95]: That `?` looks fragile, because it assumes that these will be the only URL params. Is that definitely always going to be true?  Is there a reason we don't just return `urlencode(next_params)` and leave it up to the caller to construct the URL? (See my other comment for a possible way to make this more robust.)

File: corehq/apps/api/resources/pagination.py
Comment (kaapstorm) [inline corehq/apps/api/resources/pagination.py:78]: So "next" is a URL, and `_get_next_url_params()` only returns URL params.

But this function gets called by `corehq.apps.api.resources.v0_5.get_datasource_data()` and it has access to `request`, which can tell you the URL that the request came in on. That'll give you what you need to construct a full URL, with the request params for the next page, and also to preserve any other (future?) params. Maybe unnecessary now, but also maybe more robust than assuming that next URL params will always be the only params.

This should point you in the right direction: https://docs.djangoproject.com/en/3.2/ref/request-response/#django.http.HttpRequest.build_absolute_uri

File: corehq/apps/api/resources/pagination.py
Comment (SmittieC) [reply corehq/apps/api/resources/pagination.py]: https://github.com/dimagi/commcare-hq/pull/33380/commits/fc4d78578c210e1bf5ec978ac414f70733218816

File: corehq/apps/api/resources/messaging_event/utils.py
Comment (kaapstorm) [inline corehq/apps/api/resources/messaging_event/utils.py:81]: I wasn't sure whether to add a comment here, but I will with a caveat: What you've written here is fine. You don't need to change anything.

My comment was going to be that lines where several things are happening, like this one, take me a couple of moments to parse. I try to write code as if I'm trying to explain what I'm doing to my future self, or to someone else. So when I think about it that way, firstly I tend to write shorter lines, and secondly the process of trying to explain what I'm doing often shows me ways of simplifying it.

If I was trying to explain this code to my future self, I might have written it like this:
```python
def get_request_params(request):
    has_cursor = False
    request_params = request.GET.dict()
    if 'cursor' in request_params:
        has_cursor = True
        b64_cursor = request_params.pop('cursor')
        cursor_str = b64decode(b64_cursor).decode('utf-8')
        cursor_params_dict = QueryDict(cursor_str).dict()
        request_params.update(cursor_params_dict)
    return CursorParams(request_params, request.domain, has_cursor)
```

**Review by kaapstorm — APPROVED** [review]

File: corehq/apps/api/resources/messaging_event/utils.py
Comment (SmittieC) [reply corehq/apps/api/resources/messaging_event/utils.py:81]: Thanks for highlighting. I'll do this refactor in a followup PR :+1:

**Comment by SmittieC** [comment]
I tested the management command on staging as well with my and the QA team's domain. [Here's](https://staging.commcarehq.org/admin/userreports/datasourceactionlog/30586/change/) the log entry if you're interested.

**Comment by SmittieC** [comment]
@calellowitz @snopoke this is tested on staging and only needs a final approval now

File: corehq/apps/userreports/tests/test_rebuild_migrate.py
Comment (snopoke) [inline corehq/apps/userreports/tests/test_rebuild_migrate.py:266]: why does this have a `remove_index` as well as an `add_index`?

**Review by snopoke — APPROVED** [review]
Few questions but overall looks good.

File: corehq/apps/api/resources/messaging_event/utils.py
Comment (SmittieC) [reply corehq/apps/api/resources/messaging_event/utils.py:83]: It would just make it easier to work with when reading the given parameters. Looking at this again, I think I should maybe revert this, since 1. the `messaging_events` also uses this and 2. the previous behaviour was to either return only the cursor, or the "other" params. I haven't tested the messaging AP, but don't want to mess with that behaviour. Let me change it back..

File: corehq/apps/api/resources/v0_5.py
Comment (SmittieC) [inline corehq/apps/api/resources/v0_5.py:1347]: The benefit from the reverted change was to return all params in `request_params`, so I'm just adding the `limit` param here (if it's not already there) so we can still enjoy a similar convenience. Could have just done this from the start I guess

File: corehq/apps/api/resources/messaging_event/utils.py
Comment (SmittieC) [reply corehq/apps/api/resources/messaging_event/utils.py]: https://github.com/dimagi/commcare-hq/pull/33380/commits/eafd81288e469724f347f317391e40284ccec6f3

**Comment by SmittieC** [comment]
Update: [Looks like](https://staging.commcarehq.org/admin/userreports/datasourceactionlog/?p=2) this is actually not working. We should not be seeing this

**Review by snopoke — APPROVED** [review]

**Review by calellowitz — APPROVED** [review]

**Pull request merged**
